storage/paths/db_index.go

```diff
@@ -0,0 +1,986 @@
+		ReadTs         sql.NullTime
+		ReadRefs       int
+		WriteTs        sql.NullTime
+	}
+	_, err := dbi.harmonyDB.BeginTransaction(ctx, func(tx *harmonydb.Tx) (commit bool, err error) {
+
+		fts := (read | write).AllSet()
+		err = dbi.harmonyDB.Select(ctx, &rows,
+			`SELECT sector_filetype, read_ts, read_refs, write_ts 
+								 FROM sectorlocation 
+								 WHERE miner_id=$1 
+								   AND sector_num=$2
+								   AND sector_filetype = ANY($3)`,
+			sector.Miner, sector.Number, fts)
+		if err != nil {
+			return false, xerrors.Errorf("StorageLock SELECT fails: %v", err)
+		}
+
+		type locks struct {
+			readTs   sql.NullTime
+			readRefs int
+			writeTs  sql.NullTime
+		}
+		lockMap := make(map[storiface.SectorFileType]locks)
+		for _, row := range rows {
+			lockMap[row.SectorFileType] = locks{
+				readTs:   row.ReadTs,
+				readRefs: row.ReadRefs,
+				writeTs:  row.WriteTs,
+			}
+		}
+
+		// Check if we can acquire write locks
+		// Conditions: No write lock or write lock is stale, No read lock or read lock is stale
+		for _, wft := range write.AllSet() {
+			if isLocked(lockMap[wft].writeTs) || isLocked(lockMap[wft].readTs) {
+				return false, xerrors.Errorf("cannot acquire writelock for sector %v filetype %d already locked", sector, wft)
+			}
+		}
+
+		// Check if we can acquire read locks
+		// Conditions: No write lock or write lock is stale
+		for _, rft := range read.AllSet() {
+			if isLocked(lockMap[rft].writeTs) {
+				return false, xerrors.Errorf("cannot acquire read lock for sector %v filetype %d already locked for writing", sector, rft)
+			}
+		}
+
+		// Acquire write locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sectorlocation
+							 	SET write_ts = NOW(), write_lock_owner = $1
+							 	WHERE miner_id=$2
+							 		AND sector_num=$3
+							 		AND sector_filetype = ANY($4)`,
+			lockUuid.String(),
+			sector.Miner,
+			sector.Number,
+			write.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("acquiring write locks for sector %v fails with err: %v", sector, err)
+		}
+
+		// Acquire read locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sectorlocation
+							 	SET read_ts = NOW(), read_refs = read_refs + 1
+							 	WHERE miner_id=$1
+							 		AND sector_num=$2
+							 		AND sector_filetype = ANY($3)`,
+			sector.Miner,
+			sector.Number,
+
+			read.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("acquiring read locks for sector %v fails with err: %v", sector, err)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (dbi *DBIndex) unlock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType, lockUuid uuid.UUID) (bool, error) {
+	if read|write == 0 {
+		return false, nil
+	}
+
+	if read|write > (1<<storiface.FileTypes)-1 {
+		return false, xerrors.Errorf("unknown file types specified")
+	}
+
+	// Relinquish write locks
+	_, err := dbi.harmonyDB.Exec(ctx,
```

Comment:
This ctx is guaranteed to be cancelled, we probably want to use a `.Background / .TODO` here or in `StorageLock` where we call this `unlock`

Reply:
done